hub-js/src/local/resolver/mutation/delete-type-instance.ts

```diff
@@ -45,40 +40,75 @@ export async function deleteTypeInstance(
             CALL {
                 WITH ti
                 WITH ti
-                MATCH (ti)-[:USES]->(others:TypeInstance)
-                WITH count(others) as othersLen
-                RETURN  othersLen > 1 as isUsed
+                MATCH (others:TypeInstance)-[:USES]->(ti)
+                RETURN  collect(others.id) as usedIds
             }
-            CALL apoc.util.validate(isUsed, apoc.convert.toJson({code: 400}), null)
+            CALL apoc.util.validate(size(usedIds) > 0, apoc.convert.toJson({ids: usedIds, code: 400}), null)
 
             WITH ti
             MATCH (ti)-[:CONTAINS]->(tirs: TypeInstanceResourceVersion)
             MATCH (ti)-[:OF_TYPE]->(typeRef: TypeInstanceTypeReference)
+            MATCH (ti)-[:STORED_IN]->(backendRef: TypeInstanceBackendReference)
             MATCH (metadata:TypeInstanceResourceVersionMetadata)<-[:DESCRIBED_BY]-(tirs)
             MATCH (tirs)-[:SPECIFIED_BY]->(spec: TypeInstanceResourceVersionSpec)
+            MATCH (spec)-[:WITH_BACKEND]->(specBackend: TypeInstanceResourceVersionSpecBackend)
+
             OPTIONAL MATCH (metadata)-[:CHARACTERIZED_BY]->(attrRef: AttributeReference)
 
-            DETACH DELETE ti, metadata, spec, tirs
+            // NOTE: Need to be preserved with 'WITH' statement, otherwise we won't be able
+            // to access node's properties after 'DETACH DELETE' statement.
+            WITH *, {id: ti.id, backend: { id: backendRef.id, context: specBackend.context, abstract: backendRef.abstract}} as out
+            DETACH DELETE ti, metadata, spec, tirs, specBackend
 
-            WITH typeRef
+            WITH *
             CALL {
               MATCH (typeRef)
               WHERE NOT (typeRef)--()
               DELETE (typeRef)
-              RETURN 'remove typeRef'
+              RETURN count([]) as _tmp0
             }
 
             WITH *
             CALL {
-              MATCH (attrRef)
+              MATCH (backendRef)
+              WHERE NOT (backendRef)--()
+              DELETE (backendRef)
+              RETURN count([]) as _tmp1
+            }
+
+            WITH *
+            CALL {
+              OPTIONAL MATCH (attrRef)
               WHERE attrRef IS NOT NULL AND NOT (attrRef)--()
               DELETE (attrRef)
-              RETURN 'remove attr'
+              RETURN count([]) as _tmp2
             }
 
-            RETURN $id`,
+            RETURN out`,
         { id: args.id, ownerID: args.ownerID || null }
       );
+
+      // NOTE: Use map to ensure that external storage is not called multiple time for the same ID
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const deleteExternally = new Map<string, any>();
+      result.records.forEach((record) => {
+        const out = record.get("out");
+
+        if (out.backend.abstract) {
+          return;
+        }
+        deleteExternally.set(out.id, out.backend);
+      });
+
+      for (const [id, backend] of deleteExternally) {
```

Comment:
Deleting locked TI with equal owner is not supported:

```graphql
mutation LockTypeInstances($typeInstanceId: ID!) {
  lockTypeInstances(
    in: {
      ids: [$typeInstanceId]
      ownerID: "namespace/name"
    }
  )
}

mutation DeleteLockedTypeInstance($typeInstanceId: ID!) {
  deleteTypeInstance(id: $typeInstanceId, ownerID: "namespace/name")
}
```

Result:
```json
{
  "errors": [
    {
      "message": "failed to delete TypeInstance with ID \"3de41478-f7ac-4ff8-8f35-8e7d9cb07a6d\": /storage_backend.StorageBackend/OnDelete FAILED_PRECONDITION: typeInstance locked: path \"/tmp/capact/3de41478-f7ac-4ff8-8f35-8e7d9cb07a6d\" contains \"locked_by\" property with value \"namespace/name\"",
      "locations": [
        {
          "line": 248,
          "column": 3
        }
      ],
      "path": [
        "deleteTypeInstance"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "Error: failed to delete TypeInstance with ID \"3de41478-f7ac-4ff8-8f35-8e7d9cb07a6d\": /storage_backend.StorageBackend/OnDelete FAILED_PRECONDITION: typeInstance locked: path \"/tmp/capact/3de41478-f7ac-4ff8-8f35-8e7d9cb07a6d\" contains \"locked_by\" property with value \"namespace/name\"",
            "    at deleteTypeInstance (/Users/pkosiec/repositories/capact/hub-js/dist/local/resolver/mutation/delete-type-instance.js:116:15)",
            "    at processTicksAndRejections (node:internal/process/task_queues:96:5)"
          ]
        }
      }
    }
  ],
  "data": null
}
```

Reply:
The issue was on storage backend side, to resolve it, I added a new `ownerID` property to `OnDelete` rpc.